src/main/scala/com/atomist/source/git/github/GitHubServices.scala

```diff
@@ -508,287 +362,236 @@ case class GitHubServices(oAuthToken: String, apiUrl: Option[String] = None)
-      case Failure(e) => throw ArtifactSourceUpdateException(s"Failed to create reaction", e)
-    }
-
-  private def listReactions(url: String, content: Option[ReactionContent] = None): Seq[Reaction] =
-    Try(Http(url)
-      .params(content.map(rc => Map("content" -> rc.toString)).getOrElse(Map.empty))
-      .headers(headers)
-      .execute(fromJson[Seq[Reaction]])
-      .throwError
-      .body) match {
-      case Success(reactions) => reactions
+                            events: Array[String]): Webhook = {
+    val data = toJson(Map("name" -> name, "config" -> Map("url" -> webhookUrl, "content_type" -> contentType), "active" -> active, "events" -> events))
+    retry("createWebhook") {
+      httpRequest[Webhook](url, Post, Some(data))
+    }
+  }
+
+  private def createReaction(opName: String, url: String, content: ReactionContent): Reaction = {
+    val data = toJson(Map("content" -> content.toString))
+    retry(opName) {
+      httpRequest[Reaction](url, Post, Some(data))
+    }
+  }
+
+  private def listReactions(url: String, content: Option[ReactionContent] = None): Seq[Reaction] = {
+    val params = content.map(rc => Map("content" -> rc.toString)).getOrElse(Map.empty) + ("per_page" -> "100")
+    paginateResults[Reaction](url, params)
+  }
+
+  private def httpRequestOption[T](url: String,
+                                   method: HttpMethod = Get,
+                                   data: Option[Array[Byte]] = None,
+                                   params: Map[String, String] = Map.empty)(implicit m: Manifest[T]): Option[T] =
+    Try(httpRequest(url, method, data, params)) match {
+      case Success(result) => Some(result)
       case Failure(e) =>
-        logger.debug(s"Failed to list reactions", e)
-        Seq.empty
+        logger.warn(e.getMessage, e)
+        None
     }
 
+  private def httpRequest[T](url: String,
+                             method: HttpMethod = Get,
+                             body: Option[Array[Byte]] = None,
+                             params: Map[String, String] = Map.empty)(implicit m: Manifest[T]): T =
+    (body match {
+      case Some(data) => method match {
+        case Post => Http(url).postData(data)
+        case Put => Http(url).put(data)
+        case Patch => Http(url).postData(data).method(method.toString)
+        case _ => Http(url).postData(data)
+      }
+      case None => Http(url).method(method.toString)
+    }).headers(headers)
+      .params(params)
+      .exec((code: Int, headers: Map[String, IndexedSeq[String]], is: InputStream) => code match {
+        case 200 | 201 => fromJson[T](is)
+        case success if 202 until 206 contains success =>
+          logger.debug(s"${headers("Status").head}")
+          ().asInstanceOf[T]
+        case 401 | 403 | 422 => throw DoNotRetryException(s"${headers("Status").head}")
+        case serverError if serverError >= 500 => throw DoNotRetryException(s"${headers("Status").head}")
```

Comment:
I would think 500 would be the quintessential error code to retry? The error is entirely on their end and we should retry to see if they get their stuff together at some point.